queue/limited_serializable.go

```diff
@@ -0,0 +1,680 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/utility"
+	"github.com/google/uuid"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/pool"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+)
+
+// limitedSizeSerializableLocal implements the amboy.RetryableQueue interface.
+// Unlike other implementations, the size of the queue is limited for both
+// pending jobs and completed jobs; this makes it possible to use these queues
+// in situations as parts of services and in longer-running contexts.
+// Furthermore, jobs in this queue must be serializable.
+//
+// Specify a job capacity when constructing the queue. In total, the queue will
+// store no more than 2x the specified capacity. For completed jobs, will store
+// no more than the specified capacity.
+type limitedSizeSerializableLocal struct {
+	pending                      chan amboy.Job
+	started                      bool
+	toDelete                     []string
+	capacity                     int
+	storage                      map[string]amboy.Job
+	scopes                       ScopeManager
+	dispatcher                   Dispatcher
+	retryHandler                 amboy.RetryHandler
+	staleRetryingMonitorInterval time.Duration
+	lifetimeCtx                  context.Context
+
+	retryingCount int
+	deletedCount  int
+	staleCount    int
+	id            string
+	runner        amboy.Runner
+	mu            sync.RWMutex
+}
+
+// NewLocalLimitedSizeSerializable constructs a local limited-size retryable
+// queue instance with the specified number of workers and maximum capacity.
+func NewLocalLimitedSizeSerializable(workers, capacity int) amboy.RetryableQueue {
```

Comment:
Should this have a default capacity?

Reply:
None of the implementations have a default capacity.